### PR TITLE
Support load json-data into Doris by RoutineLoad or StreamLoad

### PR DESCRIPTION
**Pull request by worker24h**
Doris support load json-data by RoutineLoad or StreamLoad. Ref: [https://github.com/apache/incubator-doris/pull/3230]

### CODE REVIEW

**Comment by morningman** [comment]
UT compile failed

File: docs/en/sql-reference/sql-statements/Data Manipulation/ROUTINE LOAD.md [hunk 1]
```diff
@@ -163,10 +163,21 @@ FROM data_source
...
 
         Specifies the time zone in which the job will be loaded. The default by using session variable's timezone. This parameter affects all function results related to the time zone involved in the load.
 
+    6. `format`
+
+        Specifies the format of the imported data. Support csv and json, the default is csv.
+
+    7. `jsonpaths`
+
+        Set jsonpath for precise imports, otherwise simple imports.
```
Comment (imay) [inline docs/en/sql-reference/sql-statements/Data Manipulation/ROUTINE LOAD.md:176]: what does `simple imports` mean? Should specify that `jsonpaths` is only valid when format is `json`

File: docs/en/sql-reference/sql-statements/Data Manipulation/ROUTINE LOAD.md [hunk 2]
```diff
@@ -163,10 +163,21 @@ FROM data_source
...
+
+    8. `strip_outer_array`
+        Boolean type, true to indicate that json data starts with an array object and flattens objects in the array object.
```
Comment (imay) [inline docs/en/sql-reference/sql-statements/Data Manipulation/ROUTINE LOAD.md:179]: what is the default value?

File: docs/en/sql-reference/sql-statements/Data Manipulation/STREAM LOAD.md [hunk 3]
```diff
@@ -111,6 +111,15 @@ Specifies the time zone used for this load. The default is East Eight District.
...
 Memory limit. Default is 2GB. Unit is Bytes.
 
+`format`
+Specifies the format of the imported data. Support csv and json, the default is csv.
+
+`jsonpaths`
+Set jsonpath for precise imports, otherwise simple imports.
+
+`strip_outer_array`
+Boolean type, true to indicate that json data starts with an array object and flattens objects in the array object.
```
Comment (imay) [inline docs/en/sql-reference/sql-statements/Data Manipulation/STREAM LOAD.md:121]: same with routine load

**Review by imay — CHANGES_REQUESTED** [review]

File: docs/zh-CN/sql-reference/sql-statements/Data Manipulation/ROUTINE LOAD.md [hunk 4]
```diff
@@ -137,15 +137,27 @@ under the License.
...
+
+            指定导入数据格式，默认是csv，支持json格式。
+
+        7. jsonpaths
+
+            jsonpaths: 导入json方式分为：简单模式和精准模式。如果设置了jsonpath则为精准模式导入，否则为简单模式导入，具体可参考示例。
+
+        8. strip_outer_array
+
+            布尔类型，为true表示json数据以数组对象开始且将数组对象中进行展平，默认值是false。
```
Comment (imay) [inline docs/zh-CN/sql-reference/sql-statements/Data Manipulation/ROUTINE LOAD.md:159]: 给个例子就能够让用户更好理解了，比如下面的数据，如果这个值设置为true，那么就会导入两条数据。如果设置为false，就会按照给定的jsonpaths来抽取字段进行导入。
```
[
    {"k1" : 1, "v1" : 2},
    {"k1" : 3, "v1" : 4},
]

```

File: be/src/exec/json_scanner.cpp [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
+        delete[] json_str;
+        std::stringstream str_error;
+        str_error << "Parse json data for JsonDoc is failed. code = " << _json_doc.GetParseError()
+                << ", error-info:" << rapidjson::GetParseError_En(_json_doc.GetParseError());
+        return Status::InternalError(str_error.str());
+    }
+
+    if (!_json_doc.IsArray() && _strip_outer_array) {
+        delete[] json_str;
+        return Status::InternalError("JSON ROOT node is array-object, `json_unnest` must be TRUE.");
```
Comment (morningman) [inline be/src/exec/json_scanner.cpp:251]:
```suggestion
        return Status::InternalError("JSON ROOT node is array-object, `strip_outer_array ` must be TRUE.");
```

File: be/src/exec/json_scanner.cpp [hunk 6]
```diff
@@ -0,0 +1,546 @@
...
+                    _counter->num_rows_filtered++;
+                    valid = false; // current row is invalid
+                    break;
+                }
+            } else {
+                RETURN_IF_ERROR(write_data_to_tuple(json_values, slot_descs[i], tuple, tuple_pool));
+            }
+        }
+        _next_line++;
+        if (!valid || nullcount == _jsonpaths_doc.Size()) {
```
Comment (morningman) [inline be/src/exec/json_scanner.cpp:519]: if `nullcount == _jsonpaths_doc.Size()`, the `_counter->num_rows_filtered++` should be called, and the error line should be recorded in runtime state.

File: docs/zh-CN/sql-reference/sql-statements/Data Manipulation/STREAM LOAD.md [hunk 7]
```diff
@@ -127,7 +142,35 @@ under the License.
...
+       json数据格式：
+           {"category":"C++","author":"avc","title":"C++ primer","price":895}
+         导入命令：
+           curl --location-trusted -u root  -H "label:123" -H "format: json" -T testData http://host:port/api/testDb/testTbl/_stream_load
+         为了提升吞吐量，支持一次性导入条数据，json数据格式如下：
+           {"RECORDS":[
+               {"category":"C++","author":"avc","title":"C++ primer","price":89.5},
+               {"category":"Java","author":"avc","title":"Effective Java","price":95},
+               {"category":"Linux","author":"avc","title":"Linux kernel","price":195}
+              ]}
```
Comment (imay) [inline docs/zh-CN/sql-reference/sql-statements/Data Manipulation/STREAM LOAD.md:160]: 我理解不应该是这样的格式，而应该是下面的格式:
[
               {"category":"C++","author":"avc","title":"C++ primer","price":89.5},
               {"category":"Java","author":"avc","title":"Effective Java","price":95},
               {"category":"Linux","author":"avc","title":"Linux kernel","price":195}
]

File: docs/zh-CN/sql-reference/sql-statements/Data Manipulation/STREAM LOAD.md [hunk 8]
```diff
@@ -67,13 +67,28 @@ under the License.
...
+
         exec_mem_limit: 导入内存限制。默认为 2GB。单位为字节。
 
+        format: 指定导入数据格式，默认是csv，支持json格式。
+
+        jsonpaths: 导入json方式分为：简单模式和精准模式。
+              简单模式：没有设置jsonpaths参数即为简单模式，这种模式下要求json数据是对象类型，例如：
+              {"k1":1, "k2":2, "k3":"hello"}，其中k1，k2，k3是列名字。
+
+              精准模式：用于json数据相对复杂，需要通过jsonpaths参数获取相应value。
```
Comment (imay) [inline docs/zh-CN/sql-reference/sql-statements/Data Manipulation/STREAM LOAD.md:83]: 感觉精准用在这里不太适合。这里完成的应该是匹配，映射工作吧？

File: docs/zh-CN/sql-reference/sql-statements/Data Manipulation/ROUTINE LOAD.md [hunk 9]
```diff
@@ -309,6 +321,84 @@ under the License.
...
+        "kafka_partitions" = "0,1,2",
+        "kafka_offsets" = "0,0,0"
+        );
+        支持两种json数据格式：
+      1）{"category":"a9jadhx","author":"test","price":895}
+        2）{
+            "RECORDS":[
+                {"category":"a9jadhx","author":"test","price":895},
+                {"category":"axdfa1","author":"EvelynWaugh","price":1299}
+            ]
```
Comment (imay) [inline docs/zh-CN/sql-reference/sql-statements/Data Manipulation/ROUTINE LOAD.md:349]:
```suggestion
        2）[
                {"category":"a9jadhx","author":"test","price":895},
                {"category":"axdfa1","author":"EvelynWaugh","price":1299}
            ]
```

File: docs/en/sql-reference/sql-statements/Data Manipulation/ROUTINE LOAD.md (hunk 1 above)
Comment (worker24h) [reply docs/en/sql-reference/sql-statements/Data Manipulation/ROUTINE LOAD.md]: i had add more comment

File: docs/en/sql-reference/sql-statements/Data Manipulation/ROUTINE LOAD.md (hunk 2 above)
Comment (worker24h) [reply docs/en/sql-reference/sql-statements/Data Manipulation/ROUTINE LOAD.md]: ok

File: docs/en/sql-reference/sql-statements/Data Manipulation/STREAM LOAD.md (hunk 3 above)
Comment (worker24h) [reply docs/en/sql-reference/sql-statements/Data Manipulation/STREAM LOAD.md]: ok

File: docs/zh-CN/sql-reference/sql-statements/Data Manipulation/ROUTINE LOAD.md [hunk 10]
```diff
@@ -137,15 +137,27 @@ under the License.
...
+
             指定导入作业所使用的时区。默认为使用 Session 的 timezone 参数。该参数会影响所有导入涉及的和时区有关的函数结果。
 
+        6. format
+
+            指定导入数据格式，默认是csv，支持json格式。
+
+        7. jsonpaths
+
+            jsonpaths: 导入json方式分为：简单模式和精准模式。如果设置了jsonpath则为精准模式导入，否则为简单模式导入，具体可参考示例。
```
[the remark this answers is not in the document]
Comment (worker24h) [reply docs/zh-CN/sql-reference/sql-statements/Data Manipulation/ROUTINE LOAD.md]: ok

File: docs/zh-CN/sql-reference/sql-statements/Data Manipulation/ROUTINE LOAD.md (hunk 4 above)
Comment (worker24h) [reply docs/zh-CN/sql-reference/sql-statements/Data Manipulation/ROUTINE LOAD.md:159]: yes， i changed

File: be/src/exec/json_scanner.cpp (hunk 5 above)
Comment (worker24h) [reply be/src/exec/json_scanner.cpp]: ok

File: be/src/exec/json_scanner.cpp (hunk 6 above)
Comment (worker24h) [reply be/src/exec/json_scanner.cpp]: ok

File: docs/zh-CN/sql-reference/sql-statements/Data Manipulation/STREAM LOAD.md (hunk 7 above)
Comment (worker24h) [reply docs/zh-CN/sql-reference/sql-statements/Data Manipulation/STREAM LOAD.md]: ok， i changed

File: docs/zh-CN/sql-reference/sql-statements/Data Manipulation/STREAM LOAD.md (hunk 8 above)
Comment (worker24h) [reply docs/zh-CN/sql-reference/sql-statements/Data Manipulation/STREAM LOAD.md]: ok

File: docs/zh-CN/sql-reference/sql-statements/Data Manipulation/ROUTINE LOAD.md (hunk 9 above)
Comment (worker24h) [reply docs/zh-CN/sql-reference/sql-statements/Data Manipulation/ROUTINE LOAD.md]: i changed

**Review by imay — COMMENTED** [review]
This interface LGTM

**Review by morningman — DISMISSED** [review]
LGTM

File: be/test/exec/json_scanner_test2.cpp [hunk 11]
```diff
@@ -0,0 +1,426 @@
...
+#include "runtime/exec_env.h"
+#include "runtime/runtime_state.h"
+#include "runtime/row_batch.h"
+#include "runtime/user_function_cache.h"
+#include "gen_cpp/Descriptors_types.h"
+#include "gen_cpp/PlanNodes_types.h"
+
+namespace doris {
+
+class JsonSannerTest : public testing::Test {
```
Comment (kangkaisen) [inline be/test/exec/json_scanner_test2.cpp:40]: What's the meaning of `json_scanner_test2`? Why do we both need `json_scanner_test` and `json_scanner_test2`?

File: be/test/exec/json_scanner_test.cpp [hunk 12]
```diff
@@ -0,0 +1,445 @@
...
+
+        t_desc_table.slotDescriptors.push_back(slot_desc);
+    }
+    byteOffset += 8;
+    t_desc_table.__isset.slotDescriptors = true;
+    {
+        // TTupleDescriptor dest
+        TTupleDescriptor t_tuple_desc;
+        t_tuple_desc.id = TUPLE_ID_DST;
+        t_tuple_desc.byteSize = byteOffset+8;//此处8字节为了处理null值
```
Comment (kangkaisen) [inline be/test/exec/json_scanner_test.cpp:232]: All comments should use english.

Comment (worker24h) [reply be/test/exec/json_scanner_test.cpp]: ok, i remove it

File: be/test/exec/json_scanner_test2.cpp (hunk 11 above)
Comment (worker24h) [reply be/test/exec/json_scanner_test2.cpp]: there are two senses, i had change the file name

**Review by imay — DISMISSED** [review]

**Review by morningman — CHANGES_REQUESTED** [review]
Compile failed.

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**